File: services/server/src/safe_socket/safe_socket.py
import socket

def recv_all(socket: socket.socket, size):
    buff = b""
    
    bytes_remaining = size
    while len(buff) < size:
        try: 
            chunk = socket.recv(bytes_remaining)
            
            # if chunk is 0 bytes, then the connection ended
            if not chunk:
                return b""

            buff += chunk
            bytes_remaining = size - len(buff)
        except OSError as err:
            raise err

    return buff

def send_all(socket: socket.socket, bytes):
    bytes_written = 0

    while bytes_written < len(bytes):
        try:
            n = socket.send(bytes[bytes_written:])
            bytes_written += n
        except OSError as err:
            return err

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	bytesWritten := 0
	for bytesWritten < len(bytes) {
		n, err := socket.Write(bytes[bytesWritten:])
		if err != nil {
			return err
		}

		bytesWritten += n
	}

	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	
	bytesRead := 0
	for bytesRead < size {
		n, err := socket.Read(buff[bytesRead:])
		bytesRead += n
		
		// an error occured, could be an EOF
		if err != nil {
			if bytesRead < size {
				return nil, err
			}
		}
	}

	return buff, nil
}
